Sources/performance-worker/BenchmarkService.swift

```diff
@@ -0,0 +1,161 @@
+            }
+          }
+        )
+      }
+      return [:]
+    }
+  }
+
+  /// Single-sided unbounded streaming from client to server.
+  /// The server returns a payload with the size requested by the client once the client does WritesDone.
+  func streamingFromClient(
+    request: ServerRequest.Stream<Grpc_Testing_BenchmarkService.Method.StreamingFromClient.Input>
+  ) async throws
+    -> ServerResponse.Single<Grpc_Testing_BenchmarkService.Method.StreamingFromClient.Output>
+  {
+    var responseSize = 0
+    for try await message in request.messages {
+      if message.responseStatus.isInitialized {
+        try self.checkOkStatus(message.responseStatus)
+      }
+      responseSize = Int(message.responseSize)
+    }
+
+    return ServerResponse.Single(
+      message: Grpc_Testing_BenchmarkService.Method.StreamingFromClient.Output.with {
+        $0.payload = Grpc_Testing_Payload.with {
+          $0.body = Data(count: responseSize)
+        }
+      }
+    )
+  }
+
+  /// Single-sided unbounded streaming from server to client.
+  /// The server repeatedly returns a payload with the size requested by the client.
+  func streamingFromServer(
+    request: ServerRequest.Single<Grpc_Testing_BenchmarkService.Method.StreamingFromServer.Input>
+  ) async throws
+    -> ServerResponse.Stream<Grpc_Testing_BenchmarkService.Method.StreamingFromServer.Output>
+  {
+    if request.message.responseStatus.isInitialized {
+      try self.checkOkStatus(request.message.responseStatus)
+    }
+    let response = Grpc_Testing_BenchmarkService.Method.StreamingCall.Output.with {
+      $0.payload = Grpc_Testing_Payload.with {
+        $0.body = Data(count: Int(request.message.responseSize))
+      }
+    }
+    return ServerResponse.Stream { writer in
+      while working.load(ordering: .acquiring) {
+        try await writer.write(response)
+      }
+      return [:]
+    }
+  }
+
+  /// Two-sided unbounded streaming between server to client.
+  /// Both sides send the content of their own choice to the other.
+  func streamingBothWays(
+    request: GRPCCore.ServerRequest.Stream<
+      Grpc_Testing_BenchmarkService.Method.StreamingBothWays.Input
+    >
+  ) async throws
+    -> GRPCCore.ServerResponse.Stream<Grpc_Testing_BenchmarkService.Method.StreamingBothWays.Output>
+  {
+    for try await message in request.messages {
+      if message.responseStatus.isInitialized {
+        try self.checkOkStatus(message.responseStatus)
+      }
+    }
+
+    // Always use the same canned response for bidirectional streaming.
+    // This is allowed by the spec.
+    let response = Grpc_Testing_BenchmarkService.Method.StreamingCall.Output.with {
+      $0.payload = Grpc_Testing_Payload.with {
+        $0.body = Data(count: 100)
+      }
+    }
+    return ServerResponse.Stream { writer in
+      while working.load(ordering: .acquiring) {
+        try await writer.write(response)
+      }
+      return [:]
+    }
```

Comment:
We need to read and write _at the same time_ which means we need to do the work in a task group within the body of the response stream. We can have one task for reading and one for writing. You can use an atomic bool as the stopping condition for writing (we should stop writing when all requests have been consumed).